File: RadioManager/cpp/radiomodel.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "RadioModel.h"

using namespace SDKTemplate;
using namespace Windows::UI::Core;
using namespace Windows::Foundation;
using namespace Windows::Devices::Radios;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Data;
using namespace concurrency;

RadioModel::RadioModel(Radio^ radio, UIElement^ parent)
{
    this->radio = radio;
    this->parent = parent;
    this->radio->StateChanged += ref new TypedEventHandler<Radio^, Object^>(this, &RadioModel::Radio_StateChanged);
}

void RadioModel::NotifyPropertyChanged(Platform::String^ propertyName)
{
    PropertyChanged(this, ref new PropertyChangedEventArgs(propertyName));
}

void RadioModel::SetRadioState(bool isRadioOn)
{
    RadioState radioState = isRadioOn ? RadioState::On : RadioState::Off;
    Disable();

    auto setStateOperation = this->radio->SetStateAsync(radioState);
    create_task(setStateOperation).then([this](task<RadioAccessStatus> asyncInfo)
    {
        NotifyPropertyChanged("IsRadioOn");
        Enable();
    });
}

void RadioModel::Radio_StateChanged(Radio^ sender, Object^ args)
{
    Dispatch([this]()
    {
        NotifyPropertyChanged("IsRadioOn");
    });
}

void RadioModel::Enable()
{
    IsEnabled = true;
}

void RadioModel::Disable()
{
    IsEnabled = false;
}

// Execute the passed-in function in the UI thread
template<typename Func>
void RadioModel::Dispatch(Func function)
{
    this->parent->Dispatcher->RunAsync(
        CoreDispatcherPriority::Normal,
        ref new DispatchedHandler(function));
}

File: RadioManager/cs/radiomodel.cs
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using Windows.ApplicationModel.Core;
using Windows.Devices.Radios;
using Windows.UI.Core;
using Windows.UI.Xaml;

namespace RadioManagerSample
{
    public class RadioModel : INotifyPropertyChanged
    {
        private Radio radio;
        private bool isEnabled = true;
        private UIElement parent;

        public RadioModel(Radio radio, UIElement parent)
        {
            this.radio = radio;
            this.parent = parent;
            this.radio.StateChanged += Radio_StateChanged;
        }

        private async void Radio_StateChanged(Radio sender, object args)
        {
            // The Radio StateChanged event doesn't run from the UI thread, so we must use the dispatcher
            // to run NotifyPropertyChanged
            await this.parent.Dispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
            {
                NotifyPropertyChanged("IsRadioOn");
            });
        }

        public string Name
        {
            get
            {
                return this.radio.Name;
            }
        }

        public bool IsRadioOn
        {
            get
            {
                return this.radio.State == RadioState.On;
            }
            set
            {
                SetRadioState(value);
            }
        }

        public bool IsEnabled
        {
            get
            {
                return this.isEnabled;
            }
            set
            {
                this.isEnabled = value;
                NotifyPropertyChanged();
            }
        }

        public event PropertyChangedEventHandler PropertyChanged;

        private void NotifyPropertyChanged([CallerMemberName] string propertyName = "")
        {
            if (PropertyChanged != null)
            {
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
            }
        }

        private async void SetRadioState(bool isRadioOn)
        {
            var radioState = isRadioOn ? RadioState.On : RadioState.Off;
            Disable();
            await this.radio.SetStateAsync(radioState);
            NotifyPropertyChanged("IsRadioOn");
            Enable();
        }

        private void Enable()
        {
            IsEnabled = true;
        }

        private void Disable()
        {
            IsEnabled = false;
        }
    }
}
